src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -0,0 +1,989 @@
+//! module, and a [`reader`] module. The `api` module has traits for defining sources and a
+//! timestamper, `impls` has example implementations of these traits, and [`render`] can spawn (or
+//! "render") the machinery to read from a source and write it into persistence. Module [`reader`]
+//! has code for a persist consumer, it reads and prints the data that the source pipeline writes
+//! to persistence.
+
+// TODO(aljoscha): Actually write to persistence and use persistence (or another consensus system)
+// to mint timestamp bindings.
+// TODO(aljoscha): Figure out which of the Send + Sync shenanigans are needed and/or if the
+// situation can be improved.
+
+use std::error::Error;
+use std::time::Duration;
+
+use timely::progress::Antichain;
+use tracing::error;
+
+use mz_ore::now::SYSTEM_TIME;
+
+use crate::error::LocationError;
+use crate::Client;
+use crate::Id;
+use crate::Location;
+
+use self::impls::ConsensusTimestamper;
+use self::impls::PersistConsensus;
+use self::impls::SequentialSource;
+
+/// Run this example via:
+///
+/// ```bash
+/// $ cargo test -p mz-persist-client -- persistent_source_example --nocapture --ignored
+/// ```
+#[tokio::test]
+#[ignore]
+async fn run() -> Result<(), Box<dyn Error>> {
+    mz_ore::test::init_logging_default("debug");
+
+    let batch_size = 2;
+    let timestamp_interval = Duration::from_secs(1);
+
+    // Make this artificially slow so we can eyeball the output.
+    let source_wait_time = Duration::from_secs(2);
+    let now_fn = SYSTEM_TIME.clone();
+
+    let source = SequentialSource::new(source_wait_time);
+
+    let persist = new_mem_persist().await?;
+    let bindings_id = Id::new();
+    let data_id = Id::new();
+
+    let (bindings_write, bindings_read) = persist.open(NO_TIMEOUT, bindings_id).await?;
+    let consensus = PersistConsensus::new(bindings_write, bindings_read);
+    let timestamper =
+        ConsensusTimestamper::new(now_fn.clone(), timestamp_interval.clone(), consensus);
+
+    let (data_write, data_read) = persist.open(NO_TIMEOUT, data_id).await?;
+
+    // First, render one instance of the source.
+    let source1 = source.clone();
+    let source_pipeline1 = mz_ore::task::spawn(|| "source-1", async move {
+        if let Err(e) = render::spawn_source_pipeline(
+            "source-1".to_string(),
+            source1,
+            timestamper,
+            batch_size,
+            data_write,
+        )
+        .await
+        {
+            error!("error in source: {:?}", e);
+        }
+    });
+
+    let reader_pipeline1 = mz_ore::task::spawn(|| "reader-1", async move {
+        if let Err(e) = reader::spawn_reader_pipeline(
+            "reader-1".to_string(),
+            data_read,
+            Antichain::from_elem(0),
+        )
+        .await
+        {
+            error!("error in reader: {:?}", e);
+        }
+    });
+
+    // TODO(aljoscha): We don't currently support multiple writers writing the same (definite) data
```

Comment:
confirmed offline that this is a bug in `Shard::write_batch`! feel free to fix it for me or wait for #11460 to land